internal/settings/providers/gcp/provider.go

```diff
@@ -0,0 +1,183 @@
+package gcp
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"github.com/qdm12/ddns-updater/internal/models"
+	"github.com/qdm12/ddns-updater/internal/settings/constants"
+	"github.com/qdm12/ddns-updater/internal/settings/errors"
+	"github.com/qdm12/ddns-updater/internal/settings/utils"
+	"github.com/qdm12/ddns-updater/pkg/publicip/ipversion"
+	"golang.org/x/exp/slices"
+	clouddns "google.golang.org/api/dns/v1"
+	"google.golang.org/api/googleapi"
+	"net"
+	"net/http"
+	"os"
+)
+
+type provider struct {
+	domain    string
+	host      string
+	zone      string
+	project   string
+	ipVersion ipversion.IPVersion
+}
+
+func New(data json.RawMessage, domain, host string,
+	ipVersion ipversion.IPVersion) (p *provider, err error) {
+	extraSettings := struct {
+		Project string `json:"project"`
+		Zone    string `json:"zone"`
+	}{}
+	if err := json.Unmarshal(data, &extraSettings); err != nil {
+		return nil, err
+	}
+	p = &provider{
+		domain:    domain,
+		host:      host,
+		ipVersion: ipVersion,
+		zone:      extraSettings.Zone,
+		project:   extraSettings.Project,
+	}
+	if err := p.isValid(); err != nil {
+		return nil, err
+	}
+	return p, nil
+}
+
+func (p *provider) isValid() error {
+	_, present := os.LookupEnv("GOOGLE_APPLICATION_CREDENTIALS")
+	if !present {
+		return errors.ErrEmptyCredentials
+	}
+
+	if len(p.project) == 0 {
+		return errors.ErrEmptyProject
+	}
+	if len(p.zone) == 0 {
+		return errors.ErrEmptyZoneIdentifier
+	}
+	return nil
+}
+
+func (p *provider) String() string {
+	return utils.ToString(p.domain, p.host, constants.GCP, p.ipVersion)
+}
+
+func (p *provider) Domain() string {
+	return p.domain
+}
+
+func (p *provider) Host() string {
+	return p.host
+}
+
+func (p *provider) IPVersion() ipversion.IPVersion {
+	return p.ipVersion
+}
+
+func (p *provider) Proxied() bool {
+	return false
+}
+
+func (p *provider) BuildDomainName() string {
+	return utils.BuildDomainName(p.host, p.domain)
+}
+
+func (p *provider) HTML() models.HTMLRow {
+	return models.HTMLRow{
+		Domain:    models.HTML(fmt.Sprintf("<a href=\"http://%s\">%s</a>", p.BuildDomainName(), p.BuildDomainName())),
+		Host:      models.HTML(p.Host()),
+		Provider:  "<a href=\"https://cloud.google.com/\">Google Cloud</a>",
+		IPVersion: models.HTML(p.ipVersion.String()),
+	}
+}
+
+func (p *provider) Update(ctx context.Context, _ *http.Client, ip net.IP) (newIP net.IP, err error) {
+	recordType := constants.A
+	if ip.To4() == nil {
+		recordType = constants.AAAA
+	}
+
+	dnsClient, err := clouddns.NewService(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	record, err := p.getRecord(dnsClient, recordType)
+	if err != nil && err != errors.ErrNotFound {
+		return nil, err
+	}
```

Comment:
I know this works fine, but I'm fighting hard to have the least dependency as possible.
Can these few calls be made without the sdk package?

Otherwise I'll look into it, feel free to let me know.

Reply:
@qdm12 

I can see, that the old code `dnsClient, err := clouddns.NewService(ctx)`

changed to `clouddns.NewService(ctx, option.WithHTTPClient(client), option.WithCredentialsJSON(p.credentials))`

I can't tell who made this change, but i can't authenticate properly when `option.WithHTTPClient(client)` is active.

I also searched for the documentation regarding `WithHTTPClient` option, but could not manage to find anything describing the difference. The official docs state, this is the way to properly authenticate: https://github.com/googleapis/google-cloud-go#authorization

But they don't mention the HTTP client option.

When I remove the `WithHTTPClient` option authorization is fully working, otherwise I receive this error:

```
ERROR getting record resource set: googleapi: Error 401: Request is missing required authentication credential. Expected OAuth 2 access token, login cookie or other valid authentication credential. See https://developers.google.com/identity/sign-in/web/devconsole-project.
```

I am happy to provide a PR that just removes the usage of the client, what do you think?

Reply:
Oh my god!! I cannot believe I skipped your response. Thanks for clarifying!
Indeed, reading the google sdk implementation code, using the `WithHTTPClient` cancels all other options so the authentication option no longer works. However, I re-invented the wheel in a84a1d664a02b09b9313cb13e1bf8a1acbbb73b1 and now it *should* work with the custom client and authentication (oauth2) with GCP.